distributed/comm/asyncio_tcp.py

```diff
@@ -0,0 +1,969 @@
+
+        nframes = len(frames)
+        frames_nbytes = [len(f) for f in frames]
+        # TODO: the old TCP comm included an extra `msg_nbytes` prefix that
+        # isn't really needed. We include it here for backwards compatibility,
+        # but this could be removed if we ever want to make another breaking
+        # change to the comms.
+        msg_nbytes = sum(frames_nbytes) + (nframes + 1) * 8
+        header = struct.pack(f"{nframes + 2}Q", msg_nbytes, nframes, *frames_nbytes)
+
+        if msg_nbytes < 4 * 1024:
+            # Always concatenate small messages
+            buffers = [b"".join([header, *frames])]
+        else:
+            buffers = coalesce_buffers([header, *frames])
+
+        if len(buffers) > 1:
+            self._transport.writelines(buffers)
+        else:
+            self._transport.write(buffers[0])
+
+        return msg_nbytes
+
+
+class TCP(Comm):
+    max_shard_size = dask.utils.parse_bytes(dask.config.get("distributed.comm.shard"))
+
+    def __init__(
+        self,
+        protocol,
+        local_addr: str,
+        peer_addr: str,
+        deserialize: bool = True,
+    ):
+        self._protocol = protocol
+        self._local_addr = local_addr
+        self._peer_addr = peer_addr
+        self.deserialize = deserialize
+        self._closed = False
+        super().__init__()
+
+        # setup a finalizer to close the protocol if the comm was never explicitly closed
+        self._finalizer = weakref.finalize(
+            self, self._protocol._close_from_finalizer, repr(self)
+        )
+        self._finalizer.atexit = False
+
+        # Fill in any extra info about this comm
+        self._extra_info = self._get_extra_info()
+
+    def _get_extra_info(self):
+        return {}
+
+    @property
+    def local_address(self) -> str:
+        return self._local_addr
+
+    @property
+    def peer_address(self) -> str:
+        return self._peer_addr
+
+    async def read(self, deserializers=None):
+        frames = await self._protocol.read()
+        try:
+            return await from_frames(
+                frames,
+                deserialize=self.deserialize,
+                deserializers=deserializers,
+                allow_offload=self.allow_offload,
+            )
+        except EOFError:
+            # Frames possibly garbled or truncated by communication error
+            self.abort()
+            raise CommClosedError("aborted stream on truncated data")
+
+    async def write(self, msg, serializers=None, on_error="message"):
+        frames = await to_frames(
+            msg,
+            allow_offload=self.allow_offload,
+            serializers=serializers,
+            on_error=on_error,
+            context={
+                "sender": self.local_info,
+                "recipient": self.remote_info,
+                **self.handshake_options,
+            },
+            frame_split_size=self.max_shard_size,
+        )
+        nbytes = await self._protocol.write(frames)
```

Comment:
Should we be doing any exception handling in `write` as we do in `read`?

Reply:
The exception handling in `read` has to do with catching an invalid (but properly received) message, there's no need for that in `write`.